server/handlers/authorization.go

```diff
@@ -18,8 +13,196 @@ import (
+}
+
+func NewAuthorizationHandler(
+	service svc.ServiceInterface,
+	htmlParserFunc func(w http.ResponseWriter, htmlFile string, params map[string]interface{}),
+) AuthorizationHandler {
+	return &authorizationHandlerImpl{
+		service:   service,
+		parseHTML: htmlParserFunc,
+	}
+}
+
+func (a *authorizationHandlerImpl) Authorize(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case "GET":
+		a.getAuthorizeHandler(w, r)
+	case "POST":
+		a.postAuthorizeHandler(w, r)
+	default:
+		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
+	}
+}
+
+func (a *authorizationHandlerImpl) getAuthorizeHandler(w http.ResponseWriter, r *http.Request) {
+	var (
+		clientID          = r.URL.Query().Get("client_id")
+		scopes            = r.URL.Query().Get("scope")
+		redirectURI       = r.URL.Query().Get("redirect_uri")
+		scopeDescriptions = []string{}
+		next              string
+	)
+
+	// use the default scope if none is provided
+	if scopes == "" {
+		scopes = constants.READ_USER_SCOPE
+	}
+
+	// add the scope descriptions
+	for _, scope := range strings.Split(scopes, ",") {
+		scopeDescriptions = append(scopeDescriptions, constants.ScopeDescriptions[scope])
+	}
+
+	// get the client
+	client, err := a.service.GetClient(clientID)
+	if err != nil {
+		log.Println(err)
+		// redirect to the redirect_uri with error
+		http.Redirect(w, r, "/error?opt=invalid_client", http.StatusSeeOther)
+		return
+	}
+
+	// generate the next url
+	uri, err := url.Parse("/authorize")
+	if err != nil {
+		log.Println(err)
+		http.Redirect(w, r, "/error?opt=server_error", http.StatusSeeOther)
+		return
+	}
+
+	uri.RawQuery = url.Values{
+		"client_id": {clientID},
+		"scope":     {scopes},
+	}.Encode()
+
+	next = uri.String()
+
+	// check that the user is logged in
+	token, err := r.Cookie("token")
+	if token == nil || err != nil {
+		http.Redirect(w, r, "/login?next="+next, http.StatusSeeOther)
+		return
+	}
+
+	user, err := a.service.GetUserByToken(token.Value)
+	if err != nil || user == nil {
+		http.Redirect(w, r, "/login?next="+next, http.StatusSeeOther)
+		return
+	}
+
+	// check that the redirect_uri is valid match the client's redirect_uri
+	if redirectURI != "" && client.RedirectURI != redirectURI {
+		http.Redirect(w, r, "/error?opt=redirect_uri_mismatch", http.StatusSeeOther)
+		return
+	}
+
+	a.parseHTML(w, "assets-v1/templates/src/pages/oauth_portal/authorize.html", map[string]interface{}{
+		"ClientName": client.Name,
+		"Scopes":     scopeDescriptions,
+		"Next":       next,
+	})
+}
+
+func (a *authorizationHandlerImpl) postAuthorizeHandler(w http.ResponseWriter, r *http.Request) {
+	var (
+		clientID        = r.URL.Query().Get("client_id")
+		scopes          = r.URL.Query().Get("scope")
+		returnResult, _ = strconv.ParseBool(r.URL.Query().Get("return_result"))
+	)
+
+	decision := r.FormValue("decision")
+	if decision != "allow" {
+		log.Println("User denied access")
```

Comment:
Here also

Reply:
This log statement is used for debugging purpose, since we dont use any logging library we use println instead but the behavior is same